### PR TITLE
Allow sol and lamports to be used as units

### PR DESCRIPTION
**Pull request by seanyoung**
Allow sol and lamports to be used as units
    
This removes the ether, wei, gwei, seconds, days, weeks, hours, and minutes keywords.
   
```solidity 
contract foo {
      int value = 10 sol;
}
```
The parser has been modified to parse any identifier after a number literal  as a unit, but to make the parser unambiguouos some errors cases are not handled as well, no worse than solc though. In many cases where number literals are not expected, they are no longer parsed and will result in a parse error rather than a parse tree which would never resolve anyway. For example:
    
    function foo(1 bar)
    (1 + 2) days    
    emit 1(foo);

Are all parse errors now. 
    
Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

File: src/sema/expression/literals.rs [hunk 1]
```diff
@@ -518,6 +518,53 @@ pub(super) fn struct_literal(
...
+            "seconds" => BigInt::from(1),
+            "minutes" => BigInt::from(60),
+            "hours" => BigInt::from(60 * 60),
+            "days" => BigInt::from(60 * 60 * 24),
+            "weeks" => BigInt::from(60 * 60 * 24 * 7),
+            "wei" => BigInt::from(1),
+            "gwei" => BigInt::from(10).pow(9u32),
+            "ether" => BigInt::from(10).pow(18u32),
+            "sol" => BigInt::from(10).pow(9u32),
+            "lamports" => BigInt::from(1),
```
Comment (seanyoung) [inline src/sema/expression/literals.rs:554]: @xermicus what should we do for substrate here, if anything?

Comment (xermicus) [reply src/sema/expression/literals.rs:554]: The currency type can be configured in the contracts pallet, so I don't think we should do anything for now.

Comment (seanyoung) [reply src/sema/expression/literals.rs:554]: I was worried that was going to be the answer.

**Review by xermicus — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,12 @@
 All notable changes to [Solang](https://github.com/hyperledger/solang/)
 will be documented here.
 
+## [Unreleased]
+
+### Added
+- The Solana units `sol` and `lamports` are now supported, e.g. `10 sol` and `100 lamports`.
+   [seanyoung](https://github.com/seanyoung)
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply CHANGELOG.md:9]: Good point, fixed in #1198

File: README.md [hunk 3]
```diff
@@ -155,30 +155,28 @@ Here is a brief description of what we envision for the next versions.
...
-| Tooling for calls between ink! <> solidity | In progress |
-| Support chain extensions for Substrate     | Not started |
-| Provide CLI for node interactions          | Not started |
+| Milestone                                    | Status      |
+|----------------------------------------------|-------------|
+| Specify values as "1 sol" and "1e9 lamports" | Completed   |
+| Call Solana's Rust contracts from Solidity   | Completed   |
+| Improvements in overflow checking            | Completed   |
+| Support Solana's Program Derived Addresses   | Completed   |
+| Call Solidity from Solana's Rust contracts   | Not started |
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply README.md:164]: It's not possible to consume Solang IDL to generate calls to Solidity. In Anchor, you need the program rust code in order to generate call into contracts. I guess you can manually construct calls but that's not ideal.

File: src/sema/expression/literals.rs [hunk 4]
```diff
@@ -518,6 +518,53 @@ pub(super) fn struct_literal(
...
+    unit: &Option<pt::Identifier>,
+    ns: &mut Namespace,
+    diagnostics: &mut Diagnostics,
+) -> BigInt {
+    if let Some(unit) = unit {
+        match unit.name.as_str() {
+            "wei" | "gwei" | "ether" if ns.target != crate::Target::EVM => {
+                diagnostics.push(Diagnostic::warning(
+                    *loc,
+                    format!("ethereum currency unit used while targeting {}", ns.target),
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/expression/literals.rs:532]: Good point. Fixed.

File: tests/contract_testcases/substrate/primitives/units.dot [hunk 5]
```diff
@@ -1,16 +1,26 @@
...
+	number_literal [label="int64 literal: 1000000000000000000\ntests/contract_testcases/substrate/primitives/units.sol:4:27-34"]
+	var_decl_5 [label="variable decl int64 wei\ntests/contract_testcases/substrate/primitives/units.sol:5:17-34"]
+	number_literal_6 [label="int64 literal: 1000000000\ntests/contract_testcases/substrate/primitives/units.sol:5:29-34"]
+	var_decl_7 [label="variable decl int64 dot\ntests/contract_testcases/substrate/primitives/units.sol:7:17-34"]
+	number_literal_8 [label="int64 literal: 1\ntests/contract_testcases/substrate/primitives/units.sol:7:29-34"]
+	diagnostic [label="found contract 'c'\nlevel Debug\ntests/contract_testcases/substrate/primitives/units.sol:2:9-9:10"]
+	diagnostic_11 [label="ethereum currency unit used while targeting substrate\nlevel Warning\ntests/contract_testcases/substrate/primitives/units.sol:4:27-34"]
+	diagnostic_12 [label="Solana currency unit used while targeting substrate\nlevel Warning\ntests/contract_testcases/substrate/primitives/units.sol:5:29-34"]
+	diagnostic_13 [label="unknown unit dot\nlevel Error\ntests/contract_testcases/substrate/primitives/units.sol:7:29-34"]
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply tests/contract_testcases/substrate/primitives/units.dot:13]: Yes, it certainly should. We are quoting source code after all. Fixed.

**Review by LucasSte — APPROVED** [review]

**Pull request merged**